// @meta2d/flow-diagram/src/data.js
export function flowData(pen, ctx) {
    var path = !ctx ? new Path2D() : ctx;
    var _a = pen.calculative.worldRect, x = _a.x, y = _a.y, width = _a.width, ex = _a.ex, ey = _a.ey;
    var penOffsetX = pen.offsetX;
    var offsetX = width / 7;
    if (penOffsetX > 1) {
        offsetX = penOffsetX;
    }
    else if (penOffsetX > 0) {
        offsetX = width * penOffsetX;
    }
    path.moveTo(x + offsetX, y);
    path.lineTo(ex, y);
    path.lineTo(x + width - offsetX, ey);
    path.lineTo(x, ey);
    path.closePath();
    if (path instanceof Path2D)
        return path;
}
export function flowDataAnchors(pen) {
    var points = [
        {
            x: 0.5,
            y: 0,
        },
        {
            x: 13 / 14,
            y: 0.5,
        },
        {
            x: 0.5,
            y: 1,
        },
        {
            x: 1 / 14,
            y: 0.5,
        },
    ];
    pen.anchors = points.map(function (_a, index) {
        var x = _a.x, y = _a.y;
        return {
            id: "" + index,
            penId: pen.id,
            x: x,
            y: y,
        };
    });
}
